### PR TITLE
feat: stacking library

### PR DESCRIPTION
**Pull request by yknl**
This PR adds a package to make it easier to integrate stacking. https://github.com/blockstack/stacks.js/issues/844

Usage documentation: https://github.com/blockstack/stacks.js/blob/feat/stacking/packages/stacking/README.md

Looking for input on naming for package and Stacker class.
`@stacks/stacking` or `@stacks/pox`?
`new Stacker()` or `new Pox()`?

Example:
```typescript
const { StacksTestnet } = require('@stacks/network');
const { Stacker } = require('@stacks/stacking');

const address = 'ST3XKKN4RPV69NN1PHFDNX3TYKXT7XPC4N8KC1ARH';
const network = new StacksTestnet();
const stacker = new Stacker(address, network);

// Check stacking eligibility
const stackingEligibility = await stacker.canLockStx({poxAddress, cycles});

// Lock STX
const stackingResults = await stacker.lockStx({ 
  amountMicroStx,
  poxAddress,
  cycles,
  key,
  burnBlockHeight
});
```

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] API reference/documentation update
- [ ] Other

## Does this introduce a breaking change?
No

## Are documentation updates required?
Yes

## Checklist
- [x] Code is commented where needed
- [x] Unit test coverage for new or modified code paths
- [x] `npm run test` passes
- [x] Changelog is updated
- [x] Tag 1 of @yknl or @zone117x for review

### CODE REVIEW

**Comment by hstove** [comment]
Hey @yknl , just wanted to flag that there is a test suite for stacking in the wallet, and it would be great to keep at least the unit tests in the library! The integration tests will be harder to setup on CI.

https://github.com/blockstack/stacks-wallet/blob/release/stacking/app/utils/stacking/pox.spec.ts#L78

**Comment by yknl** [comment]
Thanks for flagging, I've added new unit tests with jest to keep it consistent with the rest of the libs.

**Comment by agraebe** [comment]
so great to see this. I'd vote for `@stacks/stacking` - PoX seems a bit too broad. Could you also include a docs update with this? We have some Stacking docs that would need updates

File: packages/stacking/README.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+const address = 'ST3XKKN4RPV69NN1PHFDNX3TYKXT7XPC4N8KC1ARH';
+const network = new StacksTestnet();
+const stacker = new Stacker(address, network);
+```
+
+## Check stacking eligibility
+
+```typescript  
+const stackingEligibility = await stacker.canLockStx({poxAddress, cycles});
```
Comment (agraebe) [inline packages/stacking/README.md]: Should this be rather `isEligible`? If I recall correctly, the method doesn't check for balances of the STX account. So this could be true but locking wouldn't be enabled

File: packages/stacking/README.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+//   reason: 'ERR_STACKING_INVALID_LOCK_PERIOD',
+// }
+```
+
+## Get PoX info
+```typescript
+const poxInfo = await stacker.getPoxInfo();
+
+// poxInfo:
+// {
```
Comment (agraebe) [inline packages/stacking/README.md:79]: are we considering new types for the responses?

File: packages/stacking/README.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+## Stack STX
+```typescript
+const poxAddress = '1Xik14zRm29UsyS6DjhYg4iZeZqsDa8D3';
+const amountMicroStx = new BN(100000000000);
+const cycles = 10;
+const key = 'd48f215481c16cbe6426f8e557df9b78895661971d71735126545abddcd5377001';
+const burnBlockHeight = 2000;
+
+const stackingResults = await stacker.lockStx({
```
Comment (agraebe) [inline packages/stacking/README.md]: would it make sense to keep the method names similar to the function names in the stacking contract? so, `stackStx`?

otherwise, the different name might reason some folks to assume there might be differences between this JS method and the Clarity method?

File: packages/stacking/README.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+## Installation
+
+```
+npm install @stacks/stacking
+```
+
+## Initialization
+```typescript
+const { StacksTestnet } = require('@stacks/network');
+const { Stacker } = require('@stacks/stacking');
```
Comment (agraebe) [inline packages/stacking/README.md]: this is very likely out of scope, but it might be beneficial to think about it before we work on it: how would delegation impact the interface?

File: packages/stacking/README.md (hunk 1 above)
Comment (yknl) [reply packages/stacking/README.md]: The implementation of `canLockStx()` will fetch the account balance and use that as an input to the contract function call, so it does take that into account.

File: packages/stacking/README.md (hunk 2 above)
Comment (yknl) [reply packages/stacking/README.md:79]: Yes, there's a `PoxInfo` type that the lib exports.

File: packages/stacking/README.md (hunk 3 above)
Comment (yknl) [reply packages/stacking/README.md]: I'm certainly for maintaining consistency with function names. But I feel `lockStx` provides more context as to what actually happens to the tokens. I believe that's how it is presented in the current stacks wallet UI as well.

File: packages/stacking/README.md (hunk 4 above)
Comment (yknl) [reply packages/stacking/README.md]: Yep, I can see it as an additional method in the class for delegation.

File: packages/stacking/README.md (hunk 2 above)
Comment (agraebe) [reply packages/stacking/README.md:79]: fantastic!

File: packages/stacking/README.md (hunk 3 above)
Comment (agraebe) [reply packages/stacking/README.md]: I guess it is _not_ lock-stx in the Clarity contract because the locking is happening elsewhere? so the abstraction in our library would make sense. I'll reside with your best judgment here

File: packages/network/src/index.ts [hunk 5]
```diff
@@ -39,6 +47,8 @@ export class StacksMainnet implements StacksNetwork {
   accountEndpoint = '/v2/accounts';
   contractAbiEndpoint = '/v2/contracts/interface';
   readOnlyFunctionCallEndpoint = '/v2/contracts/call-read';
+
+  isMainnet= () => this.version === TransactionVersion.Mainnet;
```
Comment (kyranjamie) [inline packages/network/src/index.ts]: Prettier fail, would be cool to add a ci job to catch this

File: packages/stacking/README.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+## Installation
+
+```
+npm install @stacks/stacking
+```
+
+## Initialization
+```typescript
+const { StacksTestnet } = require('@stacks/network');
```
Comment (kyranjamie) [inline packages/stacking/README.md]: es imports?

File: packages/stacking/src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        return AddressHashMode.SerializeP2SH;
+      case 196:
+        return AddressHashMode.SerializeP2SH;
+      default: 
+        throw new Error('Invalid pox address version');
+    }
+  }
+}
+
+export function convertBTCAddress(btcAddress: string) {
```
Comment (kyranjamie) [inline packages/stacking/src/utils.ts]: `convertBtcAddress` ?

File: packages/stacking/README.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
+# @stacks/stacking [![npm](https://img.shields.io/npm/v/@stacks/stacking?color=red)](https://www.npmjs.com/package/@stacks/stacking)
```
Comment (agraebe) [inline packages/stacking/README.md:1]: It looks like this readme doesn't cover all available methods:
- getCoreInfo
- getPoxInfo
- getTargetBlockTime
- getAccountBalance
- getCycleDuration
- secondsUntilNextCycle
- stackingEnabledNextCycle
- hasMinimumRequiredStxAmount
- canLockStx
- lockStx
- getStatus
- modifyLockTxFee

File: packages/stacking/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+        return poxInfo.reward_cycle_length * targetBlockTime;
+      })
+  }
+
+  /**
+   * Get number of seconds until next reward cycle
+   * 
+   * @returns {Promise<number>} that resolves to a number if the operation succeeds
+   */
+  async secondsUntilNextCycle(): Promise<number> {
```
Comment (agraebe) [inline packages/stacking/src/index.ts]: suggest to rename this to `getSecondsUntilNextCycle` to be consistent

File: packages/stacking/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+        return blocksToNextCycle * cycleDuration;
+      })
+  } 
+
+  /**
+   * Check if stacking is enabled for next reward cycle
+   * 
+   * @returns {Promise<boolean>} that resolves to a bool if the operation succeeds
+   */
+  async stackingEnabledNextCycle(): Promise<boolean> {
```
Comment (agraebe) [inline packages/stacking/src/index.ts]: would suggest `isStackingEnabledNextCycle` to be consistent. I'd not use `get`, `is` indicates the response is a boolean

File: packages/stacking/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+
+        return callReadOnlyFunction({
+          network: this.network,
+          contractName,
+          contractAddress,
+          functionName: 'can-stack-stx',
+          senderAddress: this.address,
+          functionArgs: [
+              poxAddressCV,
+              uintCV(balance.toString()),
```
Comment (agraebe) [inline packages/stacking/src/index.ts:259]: shouldn't this be the balance - tx fees for stacking?

File: packages/stacking/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+  }
+
+  /**
+   * Check if account can lock stx
+   * 
+   * @param {CanLockStxOptions} options - a required lock STX options object 
+   * 
+   * @returns {Promise<StackingEligibility>} that resolves to a StackingEligibility object if the operation succeeds
+   */
+  async canLockStx({
```
Comment (agraebe) [inline packages/stacking/src/index.ts]: After thinking about "lock" vs "stack", I think it would be better to use "stack" for the following reasons:
- locking is not the end-goal of this action, it is to get the rewards through stacking
- we use "stacking" in the stacking contract
- our messaging seems to be "start stacking stacks". Sticking to the terminology allows us to remove the mental overhead required to understand that "stacking effectively locks up tokens, therefore locking is the right method to use"

I'd advocate for renaming this method to `canStack`

File: packages/stacking/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+  }
+
+  /**
+   * Generate and broadcast a stacking transaction to lock STX
+   * 
+   * @param {LockStxOptions} options - a required lock STX options object 
+   * 
+   * @returns {Promise<string>} that resolves to a broadcasted txid if the operation succeeds
+   */
+  async lockStx({
```
Comment (agraebe) [inline packages/stacking/src/index.ts]: if we'd adopt stacking vs locking, this could be simply `stack()`

File: packages/stacking/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+          firstRewardCycle: firstRewardCycle.value.toNumber(),
+          lockPeriod: lockPeriod.value.toNumber(),
+          poxAddress: {
+            version: version.buffer,
+            hashbytes: hashbytes.buffer
+          },
+          btcAddress: '',
+        }
+      } else if (responseCV.type === ClarityType.OptionalNone) {
+        throw new Error(`No stacking status found for ${this.address}`);
```
Comment (agraebe) [inline packages/stacking/src/index.ts]: with this method labeled as `getStatus`, no stacking shouldn't be an error but one of the plausible responses, right?

File: packages/stacking/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+      }
+    });
+  }
+
+  /**
+   * Adjust microstacks amount for locking after taking into account transaction fees
+   * 
+   * @returns {StacksTransaction} that resolves to a transaction object if the operation succeeds
+   */
+  modifyLockTxFee({ tx, amountMicroStx }: { tx: StacksTransaction; amountMicroStx: BN }) {
```
Comment (agraebe) [inline packages/stacking/src/index.ts:415]: can this be `setStackingTxFee`?

File: packages/stacking/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+  async stackingEnabledNextCycle(): Promise<boolean> {
+    return (await this.getPoxInfo()).rejection_votes_left_required > 0;
+  }
+
+  /**
+   * Check if account has minimum require amount of Stacks for stacking
+   * 
+   * @returns {Promise<boolean>} that resolves to a bool if the operation succeeds
+   */
+  async hasMinimumRequiredStxAmount(): Promise<boolean> {
```
Comment (agraebe) [inline packages/stacking/src/index.ts]: could we shorten this to: `hasMinimumStx` ?

File: packages/stacking/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+  cycles: number;
+  poxAddress: string;
+  amountMicroStx: BN;
+  burnBlockHeight: number;
+}
+
+export class Stacker {
+  constructor(
+    public address: string, 
+    public network: StacksNetwork = new StacksMainnet()) {}
```
Comment (agraebe) [inline packages/stacking/src/index.ts]: should this default to mainnet or rather testnet? Given that mainnet usage comes with real implications, would it maybe make sense to have the dev specify the mainnet as a type if confirmation?

File: packages/stacking/src/index.ts [hunk 18]
```diff
@@ -0,0 +1,410 @@
...
+          network: this.network,
+          contractName,
+          contractAddress,
+          functionName: 'can-stack-stx',
+          senderAddress: this.address,
+          functionArgs: [
+              poxAddressCV,
+              uintCV(balance.toString()),
+              uintCV(poxInfo.reward_cycle_id),
+              uintCV(cycles.toString())
```
Comment (agraebe) [inline packages/stacking/src/index.ts:261]: I believe there are a min and max for cycles - should we verify this before making the contract call?

File: packages/stacking/src/index.ts [hunk 19]
```diff
@@ -0,0 +1,410 @@
...
+   * @param {LockStxOptions} options - a required lock STX options object 
+   * 
+   * @returns {Promise<string>} that resolves to a broadcasted txid if the operation succeeds
+   */
+  async lockStx({
+    amountMicroStx,
+    poxAddress,
+    cycles,
+    key,
+    burnBlockHeight,
```
Comment (agraebe) [inline packages/stacking/src/index.ts:292]: do I understand correctly that this will default to the next cycle if not specified?

File: packages/stacking/src/index.ts [hunk 20]
```diff
@@ -0,0 +1,410 @@
...
+
+  /**
+   * Generate and broadcast a stacking transaction to lock STX
+   * 
+   * @param {LockStxOptions} options - a required lock STX options object 
+   * 
+   * @returns {Promise<string>} that resolves to a broadcasted txid if the operation succeeds
+   */
+  async lockStx({
+    amountMicroStx,
```
Comment (agraebe) [inline packages/stacking/src/index.ts:288]: will this be the max by default?

File: packages/stacking/src/constants.ts [hunk 21]
```diff
@@ -0,0 +1,21 @@
+export enum StackingErrors {
```
Comment (agraebe) [inline packages/stacking/src/constants.ts:1]: could we have a StackingError class with the error code, type (const), and a human-readable message? With that, developers could do `err.getMessage()` and simply display inside their app. If we just return the string, there's an opportunity for the developer to misinterpret the error string and/or misrepresent what it means to the user

File: packages/stacking/README.md [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+// 800000000000 
+```
+
+## Stack STX
+```typescript
+const poxAddress = '1Xik14zRm29UsyS6DjhYg4iZeZqsDa8D3';
+const amountMicroStx = new BN(100000000000);
+const cycles = 10;
+const key = 'd48f215481c16cbe6426f8e557df9b78895661971d71735126545abddcd5377001';
+const burnBlockHeight = 2000;
```
Comment (agraebe) [inline packages/stacking/README.md]: we should show how to get the latest block if the method doesn't get the height behind the scenes, right?

File: packages/stacking/src/index.ts (hunk 19 above)
Comment (agraebe) [reply packages/stacking/src/index.ts:292]: we'd likely also consider error checking - past or current height shouldn't be accepted.

On a different note: Because the only options are block numbers in the future, we could also think about this field as a delta to current block height (`startInXBlocks`). For instance, starting in 3 blocks would just require an integer of 3. With that, developers don't have to think about the block height

File: packages/stacking/src/index.ts [hunk 23]
```diff
@@ -0,0 +1,410 @@
...
+ * Lock stx options
+ *
+ * @param  {String} key - private key to sign transaction
+ * @param  {String} poxAddress - the reward Bitcoin address
+ * @param  {number} cycles - number of cycles to lock
+ * @param  {BigNum} amountMicroStx - number of microstacks to lock
+ * @param  {number} burnBlockHeight -the burnchain block height to begin lock
+ */
+export interface LockStxOptions {
+  key: string;
```
Comment (agraebe) [inline packages/stacking/src/index.ts]: could we be more descriptive here? something like `stackerPrivateKey`?

File: packages/stacking/src/index.ts [hunk 24]
```diff
@@ -0,0 +1,410 @@
...
+              uintCV(balance.toString()),
+              uintCV(poxInfo.reward_cycle_id),
+              uintCV(cycles.toString())
+            ],
+        });
+      })
+      .then((responseCV: ClarityValue) => {
+        if(responseCV.type === ClarityType.ResponseOk) {
+          return {
+            eligible: true
```
Comment (agraebe) [inline packages/stacking/src/index.ts:268]: This might be a bit confusing because the method name is `canLockStx`, no `isEligible`. Should we rather return something like `canLockStx` instead to be consistent?

File: packages/stacking/src/index.ts [hunk 25]
```diff
@@ -0,0 +1,410 @@
...
+  rejection_fraction: number;
+  reward_cycle_id: number;
+  reward_cycle_length: number;
+  rejection_votes_left_required: number;
+}
+
+export interface StackerInfo {
+  amountMicroStx: string;
+  firstRewardCycle: number;
+  lockPeriod: number;
```
Comment (agraebe) [inline packages/stacking/src/index.ts]: I'm assuming this method will be called to show either # of blocks until unlocked or timestamp of unlock event. Could we calculate these and return them as well for convenience? Something like:

- secondsUntilUnlock
- unlockBlockHeight

File: packages/stacking/README.md [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+  poxAddress,
+  cycles,
+  key,
+  burnBlockHeight
+});
+
+// stackingResults:
+// {
+//   txid: '0xf6e9dbf6a26c1b73a14738606cb2232375d1b440246e6bbc14a45b3a66618481',
+//   transaction: 'https://testnet-explorer.now.sh/txid/0xf6e9dbf6a26c1b73a14738606cb2232375d1b440246e6bbc14a45b3a66618481'
```
Comment (agraebe) [inline packages/stacking/README.md]: the URL needs to be updated, right? we shouldn't link to the `now.sh` url

File: packages/stacking/README.md [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+## Get account stacking status
+```typescript
+const stackingStatus = await stacker.getStatus();
+
+// stackingStatus:
+// {
+//   amount_microstx: '80000000000000',
+//   first_reward_cycle: 18,
+//   lock_period: 10,
+//   pox_address: {
```
Comment (agraebe) [inline packages/stacking/README.md]: the dev will submit a BTC address in a different format. It would be great if we could return the address in the same format as entered, so the dev wouldn't need to understand what we're decoding this to behind the scenes

File: packages/stacking/src/index.ts [hunk 28]
```diff
@@ -0,0 +1,410 @@
...
+      burnBlockHeight,
+    });
+    const tx = await makeContractCall({
+      ...txOptions,
+      senderKey: key,
+    });
+
+    const res = await broadcastTransaction(tx, txOptions.network as StacksNetwork);
+    if (typeof res === 'string') {
+      return res;
```
Comment (agraebe) [inline packages/stacking/src/index.ts:312]: shouldn't this also return the explorer URL?

File: packages/stacking/src/index.ts [hunk 29]
```diff
@@ -0,0 +1,410 @@
...
+        const someCV = responseCV as SomeCV;
+        const tupleCV: TupleCV = someCV.value as TupleCV;
+        const poxAddress: TupleCV = tupleCV.data["pox-addr"] as TupleCV;
+        const amountMicroStx: UIntCV = tupleCV.data["amount-ustx"] as UIntCV;
+        const firstRewardCycle: UIntCV = tupleCV.data["first-reward-cycle"] as UIntCV;
+        const lockPeriod: UIntCV = tupleCV.data["lock-period"] as UIntCV;
+        const version: BufferCV = poxAddress.data['version'] as BufferCV;
+        const hashbytes: BufferCV = poxAddress.data['hashbytes'] as BufferCV;
+      
+        return {
```
Comment (agraebe) [inline packages/stacking/src/index.ts:387]: would it be possible to return the txid/explorer URL for the stacking transaction that "reasoned" the status?

File: packages/stacking/README.md [hunk 30]
```diff
@@ -0,0 +1,94 @@
...
+  amountMicroStx,
+  poxAddress,
+  cycles,
+  key,
+  burnBlockHeight
+});
+
+// stackingResults:
+// {
+//   txid: '0xf6e9dbf6a26c1b73a14738606cb2232375d1b440246e6bbc14a45b3a66618481',
```
Comment (agraebe) [inline packages/stacking/README.md]: this is outside of the stacking lib, but should one of our packages (maybe `transactions`) expose a helper method to wait for a tx to complete? In the stacking tutorial, we share a `waitForTransactionSuccess` code sample. I'd assume many devs would want to implement that

File: packages/stacking/src/index.ts (hunk 17 above)
Comment (kyranjamie) [reply packages/stacking/src/index.ts]: I think there's a conversation to be had about network defaults.

Personally I'd err on the side of no default at all. Leave it to the user to be explicit every time. (To avoid "Wait, what's the default again?")

File: packages/stacking/src/utils.ts (hunk 7 above)
Comment (zone117x) [reply packages/stacking/src/utils.ts]: Had to use this function recently for testing in the sidecar. The current name is meaningless, should be something like `decodeBtcAddress` or `destructureBtcAddress`

Comment (yknl) [reply packages/stacking/src/utils.ts]: Changed to `decodeBtcAddress`

File: packages/stacking/src/index.ts (hunk 11 above)
Comment (yknl) [reply packages/stacking/src/index.ts:259]: I think rather than calculating this based on fees which would require another network call, this should be left up to the developer. Additionally this would not take into account sponsored transactions.

File: packages/stacking/src/index.ts (hunk 12 above)
Comment (yknl) [reply packages/stacking/src/index.ts]: Updated to `canStack`

File: packages/stacking/src/index.ts (hunk 14 above)
Comment (yknl) [reply packages/stacking/src/index.ts]: I think it would be easier for consumers to handle an error vs expecting a differently shaped response.

File: packages/stacking/src/index.ts (hunk 15 above)
Comment (yknl) [reply packages/stacking/src/index.ts:415]: That name seems confusing since it doesn't actually set the fee, but modifies the locked amount.

File: packages/stacking/src/index.ts (hunk 17 above)
Comment (yknl) [reply packages/stacking/src/index.ts]: Removed the default

File: packages/stacking/src/index.ts (hunk 18 above)
Comment (yknl) [reply packages/stacking/src/index.ts:261]: Since this min/max constant is set in the contract, it's probably a better idea to let the contract check this vs adding a constant in the library which may need to be updated if the contract changes.

File: packages/stacking/src/index.ts (hunk 19 above)
Comment (yknl) [reply packages/stacking/src/index.ts:292]: It would not default to any value. I think giving developers more control by setting the exact block here is better than making it slightly easier to use but taking away ability to be exact.

File: packages/stacking/src/index.ts (hunk 20 above)
Comment (yknl) [reply packages/stacking/src/index.ts:288]: It does not have a default.

File: packages/stacking/src/constants.ts (hunk 21 above)
Comment (yknl) [reply packages/stacking/src/constants.ts:1]: Added a helper function to address this.

File: packages/stacking/README.md (hunk 26 above)
Comment (yknl) [reply packages/stacking/README.md]: it shouldn't be showing the explorer URL, fixed the response.

File: packages/stacking/README.md (hunk 27 above)
Comment (yknl) [reply packages/stacking/README.md]: If you look at https://github.com/blockstack/stacks.js/pull/850/files#diff-c7ec1e9b062dbf6b31a30a695d5d88049b306aae4e07be04b6559c9832d0508dR5 I don't think we have enough information to reverse the data to the same BTC address entered.

File: packages/stacking/src/index.ts (hunk 28 above)
Comment (yknl) [reply packages/stacking/src/index.ts:312]: I think the consumer of the lib should construct the link vs in the lib

File: packages/stacking/README.md (hunk 30 above)
Comment (yknl) [reply packages/stacking/README.md]: A few reasons why I don't think that would be a good idea: 
- The time between broadcast and the transaction being completed might be a very long time on mainnet. Doesn't seem like good engineering practice to have a long running wait like that
- It feels like it should be part of the UI layer.

File: packages/stacking/src/index.ts (hunk 25 above)
Comment (yknl) [reply packages/stacking/src/index.ts]: That would take an extra network call to fetch block info in order to calculate. Probably best to let the developer do that separately if they need it.

**Comment by kyranjamie** [comment]
For the name @yknl, how about `new StackingClient()`?

**Comment by yknl** [comment]
> For the name @yknl, how about `new StackingClient()`?

I like it, what do you think @agraebe ?

File: packages/stacking/README.md (hunk 27 above)
Comment (agraebe) [reply packages/stacking/README.md]: discussed in discord, can't be done unfortunately

File: packages/stacking/src/index.ts (hunk 25 above)
Comment (agraebe) [reply packages/stacking/src/index.ts]: well, I guess the question is how would they use this method. My assumption is it'll be used to show the data. It's probably fair to say the progress or unlocking block/time would be important to show. 

could we allow an optional param to request this data, eg `getStatus(unlockDetails: boolean)`?

File: packages/stacking/src/index.ts (hunk 11 above)
Comment (agraebe) [reply packages/stacking/src/index.ts:259]: Maybe we could double click on this one. Because the developer will likely assume that a positive response means they're good to go, which wouldn't be the case _always_. 

- Curious about the sponsored txs: are they important for stacking? why would someone pay for these tx fees?
- Could we provide an optional method like `includeTxFees` to account for this not being critical?

File: packages/stacking/src/index.ts (hunk 18 above)
Comment (agraebe) [reply packages/stacking/src/index.ts:261]: you're right

File: packages/stacking/src/index.ts (hunk 19 above)
Comment (agraebe) [reply packages/stacking/src/index.ts:292]: I can see the tradeoffs of not being exact. mainly, I'm assuming that devs would want to get the latest block height right before calling this method. That could be avoided with a default and a delta, but I don't feel strongly either way. your call

File: packages/stacking/src/index.ts (hunk 25 above)
Comment (yknl) [reply packages/stacking/src/index.ts]: It would be better as a separate method then

File: packages/stacking/src/index.ts (hunk 28 above)
Comment (agraebe) [reply packages/stacking/src/index.ts:312]: why? I think we should do it in order to avoid the dev building the URLs - they'd want to use the same URL that we would provide anyways, right? One reason I see for not doing this is if we were to stay open to other explorers and not prioritizing one over the other, but that doesn't seem relevant right now

File: packages/stacking/src/index.ts (hunk 14 above)
Comment (agraebe) [reply packages/stacking/src/index.ts]: what if we were to wrap responses into something like this:

```json
{
   "status": "<active/inactive>",
   "details": {
      // status details
   }
}
```

File: packages/stacking/src/index.ts (hunk 15 above)
Comment (agraebe) [reply packages/stacking/src/index.ts:415]: right, makes sense

File: packages/stacking/src/index.ts (hunk 25 above)
Comment (agraebe) [reply packages/stacking/src/index.ts]: could be if devs would want to call that outside of the getstatus flow. either, way, I think that info should be easily available to them

File: packages/stacking/src/index.ts (hunk 28 above)
Comment (yknl) [reply packages/stacking/src/index.ts:312]: We should maintain separation of concern in the code. The URL can be easily constructed by the developer should they need it. And we shouldn't favour any explorer over another, so we shouldn't start doing that here. 

I don't think the goal should be to allow the developer to write as few lines of code as possible. And we should keep in mind good engineering practices and maintainability of code.

File: packages/stacking/src/index.ts (hunk 11 above)
Comment (yknl) [reply packages/stacking/src/index.ts:259]: It wouldn't be 100% in both cases, since fees can fluctuate and balances, minimums can change between calls. 

> Curious about the sponsored txs: are they important for stacking? why would someone pay for these tx fees?

A Stacking app or custodial service could take care of the fees for the user.

File: packages/stacking/src/index.ts (hunk 28 above)
Comment (aulneau) [reply packages/stacking/src/index.ts:312]: I agree, I think it's not the responsibility of this package to push people to the explorer, but whatever would be using this package to expose the url

Comment (agraebe) [reply packages/stacking/src/index.ts:312]: as per discussion in the meeting, we can close this out

File: packages/stacking/README.md (hunk 22 above)
Comment (yknl) [reply packages/stacking/README.md]: Yes we're showing that in the tutorial doc

File: packages/stacking/src/index.ts (hunk 23 above)
Comment (yknl) [reply packages/stacking/src/index.ts]: Changed to `privateKey`

File: packages/stacking/src/index.ts (hunk 24 above)
Comment (yknl) [reply packages/stacking/src/index.ts:268]: Updated to `canStack`

File: packages/stacking/src/index.ts (hunk 25 above)
Comment (yknl) [reply packages/stacking/src/index.ts]: the `getStatus` function is now making an additional call to the account API and returns the unlock block as part of the response
